### PR TITLE
implement location taps

### PR DESCRIPTION
**Pull request by AndrewCLu**

### CODE REVIEW

**Review by RiverRuby — APPROVED** [review]
i think it makes sense. i don't love that the chip type has so many fields, but each locationChip should also have an owner so i think it makes sense to combine them

File: apps/backend/prisma/schema.prisma [hunk 1]
```diff
@@ -84,6 +84,11 @@ model Chip {
   ownerEncryptionPublicKey String?
   ownerPsiPublicKeyLink    String?
   ownerUserData            Json? // Nullable JSONB field
+  isLocationChip           Boolean?
+  locationId               String?   @unique
+  locationName             String?
+  locationDescription      String?
+  locationData             Json? // Nullable JSONB field
```
Comment (stevenelleman) [inline apps/backend/prisma/schema.prisma:91]: Who's the owner of a location chip? Is it the admin that set it up? 

I think for this activation this makes sense, but for a future one these fields are basically repeated in the connection chip. Could be interesting to consider what a generic chip would look like for future activations.

Comment (AndrewCLu) [reply apps/backend/prisma/schema.prisma:91]: I think the owner could be the person who owns the chip but the issuer is the community that issues the chip, and these theoretically could be different

File: apps/backend/prisma/seed.ts [hunk 2]
```diff
@@ -22,7 +23,35 @@ async function main() {
...
+      const { signingKey, verifyingKey } = generateSignatureKeyPair();
+      return {
+        chipIssuer: "TESTING",
+        chipId: chip.chipId,
+        chipVariant: "NTAG212",
+        chipIsRegistered: true,
+        chipPublicKey: verifyingKey,
+        chipPrivateKey: signingKey,
+        chipTapCount: 0,
+        isLocationChip: true,
```
Comment (stevenelleman) [inline apps/backend/prisma/seed.ts:45]: could be good to pass in an option for seeding location vs user chip

File: apps/backend/src/lib/controller/chip/managed/tap/index.ts [hunk 3]
```diff
@@ -10,6 +10,7 @@ ManagedChipClient.prototype.GetTapFromChip = async function (
       this.prismaClient,
       tapParams
     );
+    console.log(tapResponse);
```
Comment (stevenelleman) [inline apps/backend/src/lib/controller/chip/managed/tap/index.ts:13]: nit

File: apps/frontend/src/pages/locations/[id].tsx [hunk 4]
```diff
@@ -84,9 +99,83 @@ const LocationTapModal: React.FC<LocationTapModalProps> = ({
...
+
+      // Compute the days of the week that the user has tapped in
+      const tapDays = computeWeeklyTapDays(
+        location.taps?.map((tap) => new Date(tap.timestamp))
+      );
+      setWeeklyTapDays(tapDays);
+
+      const savedTapInfo = await storage.loadSavedTapInfo();
+      // Delete saved tap info after fetching
+      await storage.deleteSavedTapInfo();
```
Comment (stevenelleman) [inline apps/frontend/src/pages/locations/[id].tsx:134]: Delete within the if case? Should we still delete the tap info if it's not equal to the id?

**Review by stevenelleman — APPROVED** [review]

File: apps/backend/src/lib/controller/chip/managed/tap/index.ts (hunk 3 above)
Comment (AndrewCLu) [reply apps/backend/src/lib/controller/chip/managed/tap/index.ts:13]: thx

File: apps/frontend/src/lib/storage/localStorage/user/location/index.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  const previousLocation = user.locations?.[tap.locationId];
+
+  if (
+    previousLocation &&
+    previousLocation.chipPublicKey &&
+    previousLocation.chipPublicKey !== tap.chipPublicKey
+  ) {
+    throw new Error("Cannot update location with a different chip");
```
[the remark this answers is not in the document]
Comment (AndrewCLu) [reply apps/frontend/src/lib/storage/localStorage/user/location/index.ts:30]: yup agreed, we need to add authentication for this

File: apps/frontend/src/pages/locations/[id].tsx (hunk 4 above)
Comment (AndrewCLu) [reply apps/frontend/src/pages/locations/[id].tsx:134]: yes i think we should, it shouldnt happen but i think case it does i think it makes sense to delete it

**Pull request merged**